### Fine Tuning du Roberta

In [1]:
 pip install pandas numpy scikit-learn transformers accelerate tqdm matplotlib seaborn paramiko tiktoken sentencepiece protobuf


[notice] A new release of pip is available: 23.3.2 -> 26.1.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [6]:
import os
import random
import paramiko
import tiktoken
import sentencepiece
import warnings
warnings.filterwarnings('ignore') 

import pandas as pd
import numpy as np

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score, classification_report

import torch
from torch.utils.data import Dataset, DataLoader

from transformers import RobertaTokenizer
from transformers import RobertaForMultipleChoice
from transformers import get_linear_schedule_with_warmup
from torch.optim import AdamW

from tqdm.auto import tqdm 
import matplotlib.pyplot as plt
import seaborn as sns

In [7]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Appareil utilisé pour l'entraînement : {device}")

if torch.cuda.is_available():
    print(f"Modèle du GPU : {torch.cuda.get_device_name(0)}")

Appareil utilisé pour l'entraînement : cuda
Modèle du GPU : NVIDIA GeForce RTX 3090


#### 1. Préparation des données

Avant de pouvoir entraîner notre modèle RoBERTa, nous avons dû concevoir une pipeline de préparation des données stricte. Il faut savoir que nous avons fait le choix d'une architecture en RobertaForMultipleChoice. En effet, une limite du Roberta classique est le nombre de tokens qui est limité à 512. Or, les commentaires de CMV sont longs donc afin de pouvoir au mieux ces tokens, au lieu de donner à notre modèle la concaténation de [OP + Argument 0 + Argument 1], on lui donne la paire [OP + Argument 0] et la paire [OP + Argument 1]. Ainsi, chaque paire bénéficie de 512 tokens.

Concrètement, dans cette partie, on reformate le dataset pour pouvoir passer à notre modèle les pairs. L'extraction se fait grâce à pair_id. On fait en sorte que l'ordre dans lequel les paires sont passées dépende du hasard afin de réduire le biais de position.

Concernant la taille du batch, disposant des GPU de Telecoms, nous avons fait le choix d'utiliser une taille de 8 qui permet d'éviter de rester bloquer dans des minimas locaux.

In [8]:
def format_cmv_data(raw_df):
    print("Reformatage des données en cours...")
    # Séparation des gagnants et perdants
    winners_df = raw_df[raw_df['success'] == 1].set_index('pair_id')
    losers_df = raw_df[raw_df['success'] == 0].set_index('pair_id')
    
    # Intersection pour ne garder que les paires complètes
    valid_pairs = winners_df.index.intersection(losers_df.index)
    
    formatted_data = []
    
    for p_id in valid_pairs:
        op = winners_df.loc[p_id, 'op_text']
        winning_arg = winners_df.loc[p_id, 'text']
        losing_arg = losers_df.loc[p_id, 'text']
        
        # Sécurité si doublons d'index
        if isinstance(op, pd.Series): op = op.iloc[0]
        if isinstance(winning_arg, pd.Series): winning_arg = winning_arg.iloc[0]
        if isinstance(losing_arg, pd.Series): losing_arg = losing_arg.iloc[0]
        
        # Randomisation pour éviter le biais de position
        if random.random() > 0.5:
            arg_0, arg_1 = winning_arg, losing_arg
            label = 0
        else:
            arg_0, arg_1 = losing_arg, winning_arg
            label = 1
            
        formatted_data.append({
            'pair_id': p_id,
            'op': str(op),
            'arg_0': str(arg_0),
            'arg_1': str(arg_1),
            'label': label
        })
        
    print(f"Extraction réussie : {len(formatted_data)} paires prêtes.")
    return formatted_data

In [9]:
class CMVDataset(Dataset):
    def __init__(self, data, tokenizer, max_length=512):
        self.data = data
        self.tokenizer = tokenizer
        self.max_length = max_length
        
    def __len__(self):
        return len(self.data)
        
    def __getitem__(self, idx):
        item = self.data[idx]
        
        # Tokenisation de l'Option 0 : [CLS] OP [SEP] Arg_0 [SEP]
        enc_0 = self.tokenizer(
            item['op'], item['arg_0'], 
            truncation=True, max_length=self.max_length, 
            padding='max_length', return_tensors='pt'
        )
        
        # Tokenisation de l'Option 1 : [CLS] OP [SEP] Arg_1 [SEP]
        enc_1 = self.tokenizer(
            item['op'], item['arg_1'], 
            truncation=True, max_length=self.max_length, 
            padding='max_length', return_tensors='pt'
        )
        
        # RoBERTa Multiple Choice attend une dimension (num_choices, seq_length)
        return {
            'input_ids': torch.cat([enc_0['input_ids'], enc_1['input_ids']], dim=0),
            'attention_mask': torch.cat([enc_0['attention_mask'], enc_1['attention_mask']], dim=0),
            'labels': torch.tensor(item['label'], dtype=torch.long)
        }

In [10]:
df = pd.read_csv('/home/infres/ydahasse-24/WAC.csv')

In [11]:
processed_data = format_cmv_data(df)

Reformatage des données en cours...
Extraction réussie : 3866 paires prêtes.


In [12]:
train_data, temp_data = train_test_split(processed_data, test_size=0.2, random_state=42)
val_data, test_data = train_test_split(temp_data, test_size=0.5, random_state=42)

print(f"Répartition : {len(train_data)} Train | {len(val_data)} Val | {len(test_data)} Test")

Répartition : 3092 Train | 387 Val | 387 Test


In [13]:
print("Chargement du Tokenizer RoBERTa...")
tokenizer = RobertaTokenizer.from_pretrained('roberta-base')

Chargement du Tokenizer RoBERTa...


In [14]:
train_dataset = CMVDataset(train_data, tokenizer)
val_dataset = CMVDataset(val_data, tokenizer)
test_dataset = CMVDataset(test_data, tokenizer)

In [15]:
BATCH_SIZE = 8

train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=BATCH_SIZE, shuffle=False)
test_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False)

#### Entrainement du modèle

In [16]:
# Hyperparamètres d'entraînement
EPOCHS = 4
LEARNING_RATE = 2e-5
WEIGHT_DECAY = 0.01
WARMUP_RATIO = 0.1
PATIENCE = 2 # Nombre d'époques à attendre avant l'arrêt précoce si pas d'amélioration

In [31]:
model = RobertaForMultipleChoice.from_pretrained('roberta-base')
model.to(device) 

# Configuration de l'Optimiseur
optimizer = AdamW(model.parameters(), lr=LEARNING_RATE, weight_decay=WEIGHT_DECAY)


config.json:   0%|          | 0.00/481 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/499M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

[transformers] RobertaForMultipleChoice LOAD REPORT from: roberta-base
Key                         | Status     | 
----------------------------+------------+-
lm_head.layer_norm.weight   | UNEXPECTED | 
lm_head.dense.bias          | UNEXPECTED | 
lm_head.bias                | UNEXPECTED | 
lm_head.layer_norm.bias     | UNEXPECTED | 
lm_head.dense.weight        | UNEXPECTED | 
classifier.weight           | MISSING    | 
roberta.pooler.dense.bias   | MISSING    | 
classifier.bias             | MISSING    | 
roberta.pooler.dense.weight | MISSING    | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


In [17]:
# Configuration du Scheduler 
total_steps = len(train_loader) * EPOCHS
num_warmup_steps = int(WARMUP_RATIO * total_steps)

scheduler = get_linear_schedule_with_warmup(
    optimizer,
    num_warmup_steps=num_warmup_steps,
    num_training_steps=total_steps
)

NameError: name 'optimizer' is not defined

In [35]:
# Variables pour le suivi et l'arrêt précoce
best_val_acc = 0.0
patience_counter = 0

print("Début de l'entraînement...")

for epoch in range(EPOCHS):
    # ========================================
    #               ENTRAÎNEMENT
    # ========================================
    model.train()
    total_train_loss = 0
    
    # Barre de progression pour le lot d'entraînement
    train_loop = tqdm(train_loader, desc=f"Epoch {epoch+1}/{EPOCHS} [Train]")
    
    for batch in train_loop:
        optimizer.zero_grad() # Réinitialisation des gradients
        
        # Envoi des données sur le GPU
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels = batch['labels'].to(device)
        
        # Forward pass (Prédiction)
        outputs = model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)
        loss = outputs.loss
        total_train_loss += loss.item()
        
        # Backward pass (Calcul des gradients)
        loss.backward()
        
        # Prévention de l'explosion des gradients (Gradient Clipping)
        torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
        
        # Mise à jour des poids
        optimizer.step()
        scheduler.step()
        
        # Mise à jour de la barre de progression
        train_loop.set_postfix(loss=loss.item())

    avg_train_loss = total_train_loss / len(train_loader)

    # ========================================
    #               VALIDATION
    # ========================================
    model.eval()
    val_preds, val_labels = [], []
    
    val_loop = tqdm(val_loader, desc=f"Epoch {epoch+1}/{EPOCHS} [Val]")
    
    # Pas de calcul de gradient pendant l'évaluation (économise mémoire et temps)
    with torch.no_grad():
        for batch in val_loop:
            input_ids = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            labels = batch['labels'].to(device)
            
            outputs = model(input_ids=input_ids, attention_mask=attention_mask)
            logits = outputs.logits # Scores bruts avant Softmax
            
            # L'option avec le plus grand score est notre prédiction
            preds = torch.argmax(logits, dim=1)
            
            val_preds.extend(preds.cpu().numpy())
            val_labels.extend(labels.cpu().numpy())
            
    val_acc = accuracy_score(val_labels, val_preds)
    
    print(f"\n--- Bilan de l'Époque {epoch+1} ---")
    print(f"Perte d'entraînement (Loss) : {avg_train_loss:.4f} | Précision de Validation : {val_acc:.4f}")
    
    # ========================================
    #             ARRÊT PRÉCOCE (Early Stopping)
    # ========================================
    if val_acc > best_val_acc:
        best_val_acc = val_acc
        # On sauvegarde le meilleur modèle localement
        torch.save(model.state_dict(), 'best_roberta_cmv.pt')
        print("=> Nouveau record ! Modèle sauvegardé sous 'best_roberta_cmv.pt'")
        patience_counter = 0 # On remet le compteur à zéro
    else:
        patience_counter += 1
        print(f"=> Pas d'amélioration. Patience : {patience_counter}/{PATIENCE}")
        if patience_counter >= PATIENCE:
            print("\nArrêt précoce déclenché : Le modèle a cessé de s'améliorer. Fin de l'entraînement.")
            break

Début de l'entraînement...


Epoch 1/4 [Train]:   0%|          | 0/387 [00:00<?, ?it/s]

Epoch 1/4 [Val]:   0%|          | 0/49 [00:00<?, ?it/s]


--- Bilan de l'Époque 1 ---
Perte d'entraînement (Loss) : 0.6875 | Précision de Validation : 0.6124
=> Nouveau record ! Modèle sauvegardé sous 'best_roberta_cmv.pt'


Epoch 2/4 [Train]:   0%|          | 0/387 [00:00<?, ?it/s]

Epoch 2/4 [Val]:   0%|          | 0/49 [00:00<?, ?it/s]


--- Bilan de l'Époque 2 ---
Perte d'entraînement (Loss) : 0.6507 | Précision de Validation : 0.6331
=> Nouveau record ! Modèle sauvegardé sous 'best_roberta_cmv.pt'


Epoch 3/4 [Train]:   0%|          | 0/387 [00:00<?, ?it/s]

Epoch 3/4 [Val]:   0%|          | 0/49 [00:00<?, ?it/s]


--- Bilan de l'Époque 3 ---
Perte d'entraînement (Loss) : 0.5864 | Précision de Validation : 0.6408
=> Nouveau record ! Modèle sauvegardé sous 'best_roberta_cmv.pt'


Epoch 4/4 [Train]:   0%|          | 0/387 [00:00<?, ?it/s]

Epoch 4/4 [Val]:   0%|          | 0/49 [00:00<?, ?it/s]


--- Bilan de l'Époque 4 ---
Perte d'entraînement (Loss) : 0.4652 | Précision de Validation : 0.6382
=> Pas d'amélioration. Patience : 1/2


In [18]:
def evaluate_on_test(model, test_loader, device):
    print("Chargement des meilleurs poids sauvegardés...")
    # On s'assure de charger le meilleur modèle (pas forcément celui de la dernière époque)
    model.load_state_dict(torch.load('best_roberta_cmv.pt'))
    model.eval() # Mode évaluation
    
    test_preds = []
    test_labels = []
    
    test_loop = tqdm(test_loader, desc="Évaluation Test")
    
    with torch.no_grad():
        for batch in test_loop:
            input_ids = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            labels = batch['labels'].to(device)
            
            outputs = model(input_ids=input_ids, attention_mask=attention_mask)
            
            # Récupération des prédictions
            preds = torch.argmax(outputs.logits, dim=1)
            
            test_preds.extend(preds.cpu().numpy())
            test_labels.extend(labels.cpu().numpy())
            
    # Calcul des métriques
    final_accuracy = accuracy_score(test_labels, test_preds)
    final_f1 = f1_score(test_labels, test_preds, average='macro')
    
    print("\n" + "="*40)
    print("RÉSULTATS FINAUX SUR LE JEU DE TEST")
    print("="*40)
    print(f"Précision (Accuracy) : {final_accuracy:.4f}")
    print(f"Score F1 (Macro)     : {final_f1:.4f}")
    print("\nRapport Détaillé :")
    
    # target_names correspondent à nos labels (0 = Arg_0 gagne, 1 = Arg_1 gagne)
    # Grâce à notre randomisation (anti-biais), ces deux classes doivent être équilibrées.
    print(classification_report(test_labels, test_preds, target_names=["Option 0 Gagne", "Option 1 Gagne"]))

# Lancement de l'évaluation
evaluate_on_test(model, test_loader, device)

NameError: name 'model' is not defined

##### Observations

**Performance générale.** Une accuracy de 65.89 % est cohérente avec
les résultats attendus pour RoBERTa-base sur ce dataset. Les travaux
antérieurs (Tan et al., 2016 ; Labruna et al., 2026) situent les
modèles de type BERT fine-tunés entre 63 et 70 % sur le split test
du WA corpus (807 paires), notre split réduit (387 paires) donnant
des résultats comparables.

**Équilibre des classes.** Le support est quasi-parfaitement équilibré
(193 vs 194 exemples), ce qui valide que la tâche est formulée comme
un problème pairwise symétrique. L'accuracy et le F1 macro sont
identiques, confirmant l'absence de biais vers une classe.

**Symétrie precision/recall.** Les scores precision et recall sont
proches pour les deux classes (écart max : 2 points), ce qui indique
que le modèle n'a pas de tendance systématique à sur-prédire l'une
ou l'autre. Les erreurs sont distribuées uniformément.

##### Limites identifiées

- **Fenêtre de 512 tokens.** RoBERTa-base ne peut traiter que 512
  tokens par exemple. Les arguments CMV étant souvent longs, une
  partie du contenu informatif est probablement tronquée, ce qui
  pénalise les cas où les points clés apparaissent en fin d'argument.

- **Capacité du modèle.** RoBERTa-base (125 M paramètres) reste un
  encodeur de taille modeste. La tâche de persuasion nécessite une
  compréhension fine des relations sémantiques et rhétoriques entre
  les arguments, ce qui dépasse les capacités d'un modèle de base.

- **Taille du jeu d'entraînement.** Avec ~2 746 exemples d'entraînement,
  le dataset reste petit pour le fine-tuning d'un transformer, ce qui
  augmente le risque de sur-apprentissage ou de sous-optimisation

#### 3. Modèle Deberta-v3-large

Les résultats obtenus avec RoBERTa-base constituent une baseline
solide mais restent en deçà du potentiel du matériel disponible
(GPU RTX 3080, 12 GB VRAM). Nous proposons de passer à
**DeBERTa-v3-large** (`microsoft/deberta-v3-large`) pour exploiter
cette capacité de calcul supplémentaire.

DeBERTa (He et al., 2021 ; Microsoft Research) est une architecture
transformer qui améliore BERT et RoBERTa sur deux points clés :

1. **Disentangled Attention.** Contrairement à RoBERTa qui encode
   position et contenu dans un seul vecteur, DeBERTa les sépare en
   deux matrices distinctes. Cela permet au modèle de mieux capturer
   les relations syntaxiques et sémantiques entre tokens éloignés —
   particulièrement utile pour comparer deux arguments longs.

2. **Enhanced Mask Decoder (EMD).** Un décodeur supplémentaire intègre
   l'information de position absolue lors du pré-entraînement, ce qui
   améliore la modélisation de la structure globale du texte.

In [40]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification
from torch.utils.data import DataLoader
from transformers import get_linear_schedule_with_warmup
from sklearn.metrics import accuracy_score, f1_score, classification_report
from tqdm import tqdm
import torch

In [32]:
class CMVDataset(Dataset):
    def __init__(self, data, tokenizer, max_length=1024):
        self.data = data
        self.tokenizer = tokenizer
        self.max_length = max_length

    def __len__(self):
        return len(self.data)

    def _head_tail_truncate(self, tokens, budget):
        """
        Si le texte dépasse `budget` tokens, garde la moitié du budget
        en tête et l'autre moitié en queue.
        """
        if len(tokens) <= budget:
            return tokens
        half = budget // 2
        return tokens[:half] + tokens[-half:]

    def __getitem__(self, idx):
        item = self.data[idx]

        op   = item['op']
        arg0 = item['arg_0']
        arg1 = item['arg_1']

        # ------------------------------------------------------------------
        # Tokenisation sans padding/truncation pour contrôler manuellement
        # ------------------------------------------------------------------
        tok_op   = self.tokenizer.encode(op,   add_special_tokens=False)
        tok_arg0 = self.tokenizer.encode(arg0, add_special_tokens=False)
        tok_arg1 = self.tokenizer.encode(arg1, add_special_tokens=False)

        # Tokens spéciaux : [CLS] ... [SEP] ... [SEP] ... [SEP]
        # Budget total = max_length - 4 tokens spéciaux
        special_tokens = 4
        budget = self.max_length - special_tokens

        # Répartition du budget : 1/4 pour l'OP, 3/8 pour chaque argument
        budget_op   = budget // 4
        budget_arg  = (budget - budget_op) // 2

        tok_op   = self._head_tail_truncate(tok_op,   budget_op)
        tok_arg0 = self._head_tail_truncate(tok_arg0, budget_arg)
        tok_arg1 = self._head_tail_truncate(tok_arg1, budget_arg)

        # ------------------------------------------------------------------
        # Reconstruction de la séquence :
        # [CLS] OP [SEP] Arg_0 [SEP] Arg_1 [SEP]
        # ------------------------------------------------------------------
        cls_id = self.tokenizer.cls_token_id
        sep_id = self.tokenizer.sep_token_id
        pad_id = self.tokenizer.pad_token_id

        input_ids = (
            [cls_id]
            + tok_op   + [sep_id]
            + tok_arg0 + [sep_id]
            + tok_arg1 + [sep_id]
        )

        attention_mask = [1] * len(input_ids)

        # Padding jusqu'à max_length
        pad_len = self.max_length - len(input_ids)
        input_ids      += [pad_id] * pad_len
        attention_mask += [0]      * pad_len

        return {
            'input_ids':      torch.tensor(input_ids,      dtype=torch.long),
            'attention_mask': torch.tensor(attention_mask, dtype=torch.long),
            'labels':         torch.tensor(item['label'],  dtype=torch.long)
        }

In [41]:
deberta_tokenizer = AutoTokenizer.from_pretrained('microsoft/deberta-v3-large')
 
train_dataset_deb = CMVDataset(train_data, deberta_tokenizer)
val_dataset_deb   = CMVDataset(val_data,   deberta_tokenizer)
test_dataset_deb  = CMVDataset(test_data,  deberta_tokenizer)

In [68]:
MICRO_BATCH_SIZE   = 1
ACCUMULATION_STEPS = 4
 
train_loader_deb = DataLoader(train_dataset_deb, batch_size=MICRO_BATCH_SIZE, shuffle=True)
val_loader_deb   = DataLoader(val_dataset_deb,   batch_size=MICRO_BATCH_SIZE, shuffle=False)
test_loader_deb  = DataLoader(test_dataset_deb,  batch_size=MICRO_BATCH_SIZE, shuffle=False)

In [55]:
model_deb = AutoModelForSequenceClassification.from_pretrained(
    'microsoft/deberta-v3-large',
    num_labels=2,
    torch_dtype=torch.bfloat16  
).to(device)


Loading weights:   0%|          | 0/390 [00:00<?, ?it/s]

[transformers] DebertaV2ForSequenceClassification LOAD REPORT from: microsoft/deberta-v3-large
Key                                     | Status     | 
----------------------------------------+------------+-
mask_predictions.classifier.bias        | UNEXPECTED | 
lm_predictions.lm_head.dense.weight     | UNEXPECTED | 
mask_predictions.dense.weight           | UNEXPECTED | 
lm_predictions.lm_head.dense.bias       | UNEXPECTED | 
mask_predictions.LayerNorm.weight       | UNEXPECTED | 
mask_predictions.classifier.weight      | UNEXPECTED | 
lm_predictions.lm_head.LayerNorm.weight | UNEXPECTED | 
lm_predictions.lm_head.bias             | UNEXPECTED | 
mask_predictions.LayerNorm.bias         | UNEXPECTED | 
mask_predictions.dense.bias             | UNEXPECTED | 
lm_predictions.lm_head.LayerNorm.bias   | UNEXPECTED | 
classifier.bias                         | MISSING    | 
pooler.dense.bias                       | MISSING    | 
pooler.dense.weight                     | MISSING    | 
classifie

In [69]:
LEARNING_RATE  = 1e-5
EPOCHS         = 3 

total_steps = (len(train_loader_deb) // ACCUMULATION_STEPS) * EPOCHS
scheduler = get_linear_schedule_with_warmup(
    optimizer,
    num_warmup_steps=int(0.1 * total_steps),
    num_training_steps=total_steps
)

In [70]:
best_val_acc = 0.0
print("Début de l'entraînement DeBERTa-v3-large...\n")

for epoch in range(EPOCHS):
    model_deb.train()
    total_train_loss = 0.0
    optimizer.zero_grad()

    train_loop = tqdm(train_loader_deb, desc=f"Epoch {epoch+1}/{EPOCHS} [Train]")

    for step, batch in enumerate(train_loop):
        input_ids      = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels         = batch['labels'].to(device)

        outputs = model_deb(
            input_ids=input_ids,
            attention_mask=attention_mask,
            labels=labels
        )
        loss = outputs.loss / ACCUMULATION_STEPS
        loss.backward()
        total_train_loss += outputs.loss.item()

        # Print tous les 50 steps — après le forward, pas avant
        if step % 50 == 0:
            print(f"Step {step} | loss : {outputs.loss.item():.4f}")

        if (step + 1) % ACCUMULATION_STEPS == 0 or (step + 1) == len(train_loader_deb):
            torch.nn.utils.clip_grad_norm_(model_deb.parameters(), 1.0)
            optimizer.step()
            scheduler.step()
            optimizer.zero_grad()

        train_loop.set_postfix(loss=f"{outputs.loss.item():.4f}")

    avg_train_loss = total_train_loss / len(train_loader_deb)

    # VALIDATION
    model_deb.eval()
    val_preds, val_labels_list = [], []

    with torch.no_grad():
        for batch in tqdm(val_loader_deb, desc=f"Epoch {epoch+1}/{EPOCHS} [Val]"):
            input_ids      = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            labels         = batch['labels'].to(device)

            outputs = model_deb(input_ids=input_ids, attention_mask=attention_mask)
            preds = torch.argmax(outputs.logits, dim=1)
            val_preds.extend(preds.cpu().numpy())
            val_labels_list.extend(labels.cpu().numpy())

    val_acc = accuracy_score(val_labels_list, val_preds)
    val_f1  = f1_score(val_labels_list, val_preds, average='macro')

    print(f"\nBilan Epoch {epoch+1} | Loss train : {avg_train_loss:.4f} | Val Acc : {val_acc:.4f} | Val F1 : {val_f1:.4f}")

    if val_acc > best_val_acc:
        best_val_acc = val_acc
        torch.save(model_deb.state_dict(), 'best_deberta_cmv.pt')
        print("  => Meilleur modèle sauvegardé (best_deberta_cmv.pt)")

Début de l'entraînement DeBERTa-v3-large...



Epoch 1/3 [Train]:   0%|                                   | 1/3092 [00:00<11:53,  4.33it/s, loss=0.7421]

Step 0 | loss : 0.7421


Epoch 1/3 [Train]:   2%|▌                                 | 51/3092 [00:11<11:06,  4.56it/s, loss=0.7289]

Step 50 | loss : 0.7289


Epoch 1/3 [Train]:   3%|█                                | 101/3092 [00:22<10:22,  4.80it/s, loss=0.6530]

Step 100 | loss : 0.6530


Epoch 1/3 [Train]:   5%|█▌                               | 151/3092 [00:32<11:16,  4.35it/s, loss=0.7691]

Step 150 | loss : 0.7691


Epoch 1/3 [Train]:   7%|██▏                              | 201/3092 [00:43<11:08,  4.33it/s, loss=0.6516]

Step 200 | loss : 0.6516


Epoch 1/3 [Train]:   8%|██▋                              | 251/3092 [00:54<11:40,  4.05it/s, loss=0.6748]

Step 250 | loss : 0.6748


Epoch 1/3 [Train]:  10%|███▏                             | 301/3092 [01:04<09:38,  4.82it/s, loss=0.6199]

Step 300 | loss : 0.6199


Epoch 1/3 [Train]:  11%|███▋                             | 351/3092 [01:15<08:44,  5.23it/s, loss=0.7637]

Step 350 | loss : 0.7637


Epoch 1/3 [Train]:  13%|████▎                            | 402/3092 [01:26<09:59,  4.49it/s, loss=0.7207]

Step 400 | loss : 0.8015


Epoch 1/3 [Train]:  15%|████▊                            | 451/3092 [01:37<08:41,  5.07it/s, loss=0.4410]

Step 450 | loss : 0.4410


Epoch 1/3 [Train]:  16%|█████▎                           | 502/3092 [01:48<08:31,  5.06it/s, loss=0.6206]

Step 500 | loss : 0.6717


Epoch 1/3 [Train]:  18%|█████▉                           | 551/3092 [01:58<08:45,  4.84it/s, loss=0.7793]

Step 550 | loss : 0.7793


Epoch 1/3 [Train]:  19%|██████▍                          | 602/3092 [02:10<09:01,  4.60it/s, loss=0.7380]

Step 600 | loss : 0.7680


Epoch 1/3 [Train]:  21%|██████▉                          | 651/3092 [02:20<07:57,  5.11it/s, loss=0.5108]

Step 650 | loss : 0.5108


Epoch 1/3 [Train]:  23%|███████▍                         | 702/3092 [02:31<08:05,  4.92it/s, loss=0.8481]

Step 700 | loss : 0.8632


Epoch 1/3 [Train]:  24%|████████                         | 751/3092 [02:42<08:19,  4.68it/s, loss=0.8621]

Step 750 | loss : 0.8621


Epoch 1/3 [Train]:  26%|████████▌                        | 801/3092 [02:53<08:14,  4.63it/s, loss=0.6746]

Step 800 | loss : 0.6746


Epoch 1/3 [Train]:  28%|█████████                        | 851/3092 [03:04<08:06,  4.61it/s, loss=0.6258]

Step 850 | loss : 0.6258


Epoch 1/3 [Train]:  29%|█████████▌                       | 901/3092 [03:15<08:18,  4.40it/s, loss=0.9086]

Step 900 | loss : 0.9086


Epoch 1/3 [Train]:  31%|██████████▏                      | 951/3092 [03:27<08:45,  4.07it/s, loss=0.6890]

Step 950 | loss : 0.6890


Epoch 1/3 [Train]:  32%|██████████▎                     | 1002/3092 [03:38<07:00,  4.97it/s, loss=0.6992]

Step 1000 | loss : 0.7161


Epoch 1/3 [Train]:  34%|██████████▉                     | 1051/3092 [03:48<06:51,  4.96it/s, loss=0.6499]

Step 1050 | loss : 0.6499


Epoch 1/3 [Train]:  36%|███████████▍                    | 1101/3092 [03:59<07:29,  4.43it/s, loss=0.6383]

Step 1100 | loss : 0.6383


Epoch 1/3 [Train]:  37%|███████████▉                    | 1151/3092 [04:10<06:52,  4.70it/s, loss=0.7099]

Step 1150 | loss : 0.7099


Epoch 1/3 [Train]:  39%|████████████▍                   | 1201/3092 [04:22<06:38,  4.74it/s, loss=0.7259]

Step 1200 | loss : 0.7259


Epoch 1/3 [Train]:  40%|████████████▉                   | 1251/3092 [04:33<06:25,  4.78it/s, loss=0.6262]

Step 1250 | loss : 0.6262


Epoch 1/3 [Train]:  42%|█████████████▍                  | 1301/3092 [04:44<06:28,  4.61it/s, loss=0.5082]

Step 1300 | loss : 0.5082


Epoch 1/3 [Train]:  44%|█████████████▉                  | 1351/3092 [04:55<06:12,  4.67it/s, loss=0.8235]

Step 1350 | loss : 0.8235


Epoch 1/3 [Train]:  45%|██████████████▍                 | 1401/3092 [05:07<06:50,  4.12it/s, loss=0.5843]

Step 1400 | loss : 0.5843


Epoch 1/3 [Train]:  47%|███████████████                 | 1451/3092 [05:18<06:12,  4.40it/s, loss=0.6871]

Step 1450 | loss : 0.6871


Epoch 1/3 [Train]:  49%|███████████████▌                | 1502/3092 [05:29<05:27,  4.85it/s, loss=1.1322]

Step 1500 | loss : 0.6294


Epoch 1/3 [Train]:  50%|████████████████                | 1551/3092 [05:40<05:42,  4.50it/s, loss=0.6418]

Step 1550 | loss : 0.6418


Epoch 1/3 [Train]:  52%|████████████████▌               | 1601/3092 [05:51<05:56,  4.18it/s, loss=0.6427]

Step 1600 | loss : 0.6427


Epoch 1/3 [Train]:  53%|█████████████████               | 1651/3092 [06:01<05:06,  4.69it/s, loss=0.8034]

Step 1650 | loss : 0.8034


Epoch 1/3 [Train]:  55%|█████████████████▌              | 1701/3092 [06:12<04:53,  4.74it/s, loss=0.7315]

Step 1700 | loss : 0.7315


Epoch 1/3 [Train]:  57%|██████████████████              | 1751/3092 [06:23<04:33,  4.90it/s, loss=0.5989]

Step 1750 | loss : 0.5989


Epoch 1/3 [Train]:  58%|██████████████████▋             | 1802/3092 [06:35<04:45,  4.51it/s, loss=0.6359]

Step 1800 | loss : 0.6612


Epoch 1/3 [Train]:  60%|███████████████████▏            | 1851/3092 [06:46<04:53,  4.23it/s, loss=0.6875]

Step 1850 | loss : 0.6875


Epoch 1/3 [Train]:  62%|███████████████████▋            | 1902/3092 [06:57<03:56,  5.03it/s, loss=0.6791]

Step 1900 | loss : 0.6981


Epoch 1/3 [Train]:  63%|████████████████████▏           | 1951/3092 [07:08<04:31,  4.21it/s, loss=0.6614]

Step 1950 | loss : 0.6614


Epoch 1/3 [Train]:  65%|████████████████████▋           | 2001/3092 [07:19<04:16,  4.25it/s, loss=0.7817]

Step 2000 | loss : 0.7817


Epoch 1/3 [Train]:  66%|█████████████████████▏          | 2051/3092 [07:31<04:01,  4.31it/s, loss=0.7365]

Step 2050 | loss : 0.7365


Epoch 1/3 [Train]:  68%|█████████████████████▋          | 2101/3092 [07:42<03:34,  4.61it/s, loss=0.6709]

Step 2100 | loss : 0.6709


Epoch 1/3 [Train]:  70%|██████████████████████▎         | 2151/3092 [07:53<03:28,  4.52it/s, loss=0.7124]

Step 2150 | loss : 0.6849


Epoch 1/3 [Train]:  71%|██████████████████████▊         | 2201/3092 [08:04<03:31,  4.22it/s, loss=0.5272]

Step 2200 | loss : 0.5272


Epoch 1/3 [Train]:  73%|███████████████████████▎        | 2251/3092 [08:15<03:16,  4.28it/s, loss=0.7152]

Step 2250 | loss : 0.7152


Epoch 1/3 [Train]:  74%|███████████████████████▊        | 2301/3092 [08:26<02:39,  4.95it/s, loss=0.7114]

Step 2300 | loss : 0.7114


Epoch 1/3 [Train]:  76%|████████████████████████▎       | 2351/3092 [08:37<02:27,  5.02it/s, loss=0.7483]

Step 2350 | loss : 0.7483


Epoch 1/3 [Train]:  78%|████████████████████████▊       | 2401/3092 [08:48<02:36,  4.42it/s, loss=0.6457]

Step 2400 | loss : 0.6457


Epoch 1/3 [Train]:  79%|█████████████████████████▎      | 2451/3092 [09:00<02:30,  4.25it/s, loss=0.5939]

Step 2450 | loss : 0.5939


Epoch 1/3 [Train]:  81%|█████████████████████████▉      | 2501/3092 [09:11<02:10,  4.53it/s, loss=0.6764]

Step 2500 | loss : 0.6764


Epoch 1/3 [Train]:  83%|██████████████████████████▍     | 2551/3092 [09:22<01:58,  4.57it/s, loss=0.7514]

Step 2550 | loss : 0.7514


Epoch 1/3 [Train]:  84%|██████████████████████████▉     | 2602/3092 [09:33<01:53,  4.33it/s, loss=0.5666]

Step 2600 | loss : 0.6367


Epoch 1/3 [Train]:  86%|███████████████████████████▍    | 2651/3092 [09:44<01:35,  4.64it/s, loss=0.8545]

Step 2650 | loss : 0.8545


Epoch 1/3 [Train]:  87%|███████████████████████████▉    | 2701/3092 [09:56<01:18,  4.97it/s, loss=0.7335]

Step 2700 | loss : 0.7335


Epoch 1/3 [Train]:  89%|████████████████████████████▍   | 2751/3092 [10:07<01:18,  4.34it/s, loss=0.7822]

Step 2750 | loss : 0.7822


Epoch 1/3 [Train]:  91%|████████████████████████████▉   | 2802/3092 [10:18<01:01,  4.71it/s, loss=0.8492]

Step 2800 | loss : 0.5390


Epoch 1/3 [Train]:  92%|█████████████████████████████▌  | 2851/3092 [10:29<00:53,  4.51it/s, loss=0.8262]

Step 2850 | loss : 0.8262


Epoch 1/3 [Train]:  94%|██████████████████████████████  | 2901/3092 [10:40<00:43,  4.37it/s, loss=0.8573]

Step 2900 | loss : 0.8573


Epoch 1/3 [Train]:  95%|██████████████████████████████▌ | 2951/3092 [10:51<00:30,  4.67it/s, loss=0.6453]

Step 2950 | loss : 0.6453


Epoch 1/3 [Train]:  97%|███████████████████████████████ | 3002/3092 [11:02<00:20,  4.50it/s, loss=0.5679]

Step 3000 | loss : 0.7817


Epoch 1/3 [Train]:  99%|███████████████████████████████▌| 3051/3092 [11:13<00:09,  4.30it/s, loss=0.6931]

Step 3050 | loss : 0.6931


Epoch 1/3 [Val]: 100%|█████████████████████████████████████████████████| 387/387 [00:24<00:00, 16.00it/s]



Bilan Epoch 1 | Loss train : 0.6923 | Val Acc : 0.5866 | Val F1 : 0.5861
  => Meilleur modèle sauvegardé (best_deberta_cmv.pt)


Epoch 2/3 [Train]:   0%|                                   | 2/3092 [00:00<11:17,  4.56it/s, loss=0.6546]

Step 0 | loss : 0.6405


Epoch 2/3 [Train]:   2%|▌                                 | 51/3092 [00:11<09:56,  5.10it/s, loss=0.5843]

Step 50 | loss : 0.5843


Epoch 2/3 [Train]:   3%|█                                | 101/3092 [00:21<11:21,  4.39it/s, loss=0.7224]

Step 100 | loss : 0.7224


Epoch 2/3 [Train]:   5%|█▌                               | 151/3092 [00:32<09:58,  4.91it/s, loss=0.6939]

Step 150 | loss : 0.6939


Epoch 2/3 [Train]:   7%|██▏                              | 202/3092 [00:43<09:57,  4.84it/s, loss=1.0906]

Step 200 | loss : 0.6897


Epoch 2/3 [Train]:   8%|██▋                              | 251/3092 [00:54<09:38,  4.91it/s, loss=0.6497]

Step 250 | loss : 0.6497


Epoch 2/3 [Train]:  10%|███▏                             | 302/3092 [01:06<10:16,  4.52it/s, loss=0.6558]

Step 300 | loss : 0.4804


Epoch 2/3 [Train]:  11%|███▋                             | 351/3092 [01:16<09:30,  4.81it/s, loss=0.6172]

Step 350 | loss : 0.6172


Epoch 2/3 [Train]:  13%|████▎                            | 401/3092 [01:27<09:33,  4.69it/s, loss=0.7473]

Step 400 | loss : 0.7473


Epoch 2/3 [Train]:  15%|████▊                            | 451/3092 [01:39<09:20,  4.71it/s, loss=0.5615]

Step 450 | loss : 0.5615


Epoch 2/3 [Train]:  16%|█████▎                           | 502/3092 [01:49<08:40,  4.97it/s, loss=0.8666]

Step 500 | loss : 0.5299


Epoch 2/3 [Train]:  18%|█████▉                           | 551/3092 [02:00<09:05,  4.66it/s, loss=0.6502]

Step 550 | loss : 0.6502


Epoch 2/3 [Train]:  19%|██████▍                          | 602/3092 [02:11<08:31,  4.87it/s, loss=0.8837]

Step 600 | loss : 0.6713


Epoch 2/3 [Train]:  21%|██████▉                          | 651/3092 [02:22<08:41,  4.68it/s, loss=0.7249]

Step 650 | loss : 0.7249


Epoch 2/3 [Train]:  23%|███████▍                         | 702/3092 [02:33<07:55,  5.02it/s, loss=0.9857]

Step 700 | loss : 0.6614


Epoch 2/3 [Train]:  24%|████████                         | 752/3092 [02:44<07:51,  4.96it/s, loss=0.5681]

Step 750 | loss : 0.5213


Epoch 2/3 [Train]:  26%|████████▌                        | 802/3092 [02:55<07:54,  4.83it/s, loss=0.5696]

Step 800 | loss : 0.4850


Epoch 2/3 [Train]:  28%|█████████                        | 851/3092 [03:06<07:59,  4.67it/s, loss=0.5016]

Step 850 | loss : 0.5016


Epoch 2/3 [Train]:  29%|█████████▌                       | 901/3092 [03:17<07:18,  4.99it/s, loss=0.6222]

Step 900 | loss : 0.6222


Epoch 2/3 [Train]:  31%|██████████▏                      | 951/3092 [03:28<07:17,  4.89it/s, loss=0.7269]

Step 950 | loss : 0.7269


Epoch 2/3 [Train]:  32%|██████████▎                     | 1002/3092 [03:39<07:26,  4.68it/s, loss=0.8695]

Step 1000 | loss : 0.5219


Epoch 2/3 [Train]:  34%|██████████▉                     | 1051/3092 [03:50<07:58,  4.27it/s, loss=0.6221]

Step 1050 | loss : 0.6221


Epoch 2/3 [Train]:  36%|███████████▍                    | 1102/3092 [04:01<06:44,  4.92it/s, loss=0.9033]

Step 1100 | loss : 0.9150


Epoch 2/3 [Train]:  37%|███████████▉                    | 1151/3092 [04:11<06:46,  4.78it/s, loss=0.5264]

Step 1150 | loss : 0.5264


Epoch 2/3 [Train]:  39%|████████████▍                   | 1201/3092 [04:22<06:40,  4.72it/s, loss=0.7464]

Step 1200 | loss : 0.7464


Epoch 2/3 [Train]:  40%|████████████▉                   | 1251/3092 [04:34<07:17,  4.21it/s, loss=0.8481]

Step 1250 | loss : 0.8481


Epoch 2/3 [Train]:  42%|█████████████▍                  | 1301/3092 [04:45<06:54,  4.32it/s, loss=0.8958]

Step 1300 | loss : 0.8958


Epoch 2/3 [Train]:  44%|█████████████▉                  | 1351/3092 [04:56<06:17,  4.62it/s, loss=0.8292]

Step 1350 | loss : 0.8292


Epoch 2/3 [Train]:  45%|██████████████▌                 | 1402/3092 [05:07<06:45,  4.16it/s, loss=0.5956]

Step 1400 | loss : 0.6912


Epoch 2/3 [Train]:  47%|███████████████                 | 1451/3092 [05:18<06:21,  4.31it/s, loss=0.6719]

Step 1450 | loss : 0.6719


Epoch 2/3 [Train]:  49%|███████████████▌                | 1502/3092 [05:29<05:59,  4.43it/s, loss=0.5655]

Step 1500 | loss : 0.5876


Epoch 2/3 [Train]:  50%|████████████████                | 1551/3092 [05:40<05:49,  4.41it/s, loss=0.6801]

Step 1550 | loss : 0.6801


Epoch 2/3 [Train]:  52%|████████████████▌               | 1602/3092 [05:51<04:45,  5.22it/s, loss=0.6927]

Step 1600 | loss : 0.9429


Epoch 2/3 [Train]:  53%|█████████████████               | 1651/3092 [06:01<04:51,  4.94it/s, loss=0.6091]

Step 1650 | loss : 0.6091


Epoch 2/3 [Train]:  55%|█████████████████▌              | 1702/3092 [06:13<04:33,  5.09it/s, loss=0.5001]

Step 1700 | loss : 0.5354


Epoch 2/3 [Train]:  57%|██████████████████              | 1751/3092 [06:24<05:36,  3.99it/s, loss=0.9097]

Step 1750 | loss : 0.9097


Epoch 2/3 [Train]:  58%|██████████████████▋             | 1801/3092 [06:35<04:22,  4.93it/s, loss=0.7044]

Step 1800 | loss : 0.7044


Epoch 2/3 [Train]:  60%|███████████████████▏            | 1851/3092 [06:46<04:34,  4.53it/s, loss=0.7988]

Step 1850 | loss : 0.7988


Epoch 2/3 [Train]:  62%|███████████████████▋            | 1902/3092 [06:58<04:35,  4.33it/s, loss=0.7911]

Step 1900 | loss : 0.7483


Epoch 2/3 [Train]:  63%|████████████████████▏           | 1951/3092 [07:08<04:40,  4.07it/s, loss=0.6056]

Step 1950 | loss : 0.6056


Epoch 2/3 [Train]:  65%|████████████████████▋           | 2001/3092 [07:19<03:42,  4.90it/s, loss=0.5717]

Step 2000 | loss : 0.5717


Epoch 2/3 [Train]:  66%|█████████████████████▏          | 2051/3092 [07:31<03:51,  4.50it/s, loss=0.7654]

Step 2050 | loss : 0.7654


Epoch 2/3 [Train]:  68%|█████████████████████▋          | 2101/3092 [07:41<03:18,  5.00it/s, loss=0.6797]

Step 2100 | loss : 0.6797


Epoch 2/3 [Train]:  70%|██████████████████████▎         | 2151/3092 [07:52<03:05,  5.08it/s, loss=0.5662]

Step 2150 | loss : 0.5662


Epoch 2/3 [Train]:  71%|██████████████████████▊         | 2202/3092 [08:03<02:55,  5.06it/s, loss=0.8752]

Step 2200 | loss : 0.7953


Epoch 2/3 [Train]:  73%|███████████████████████▎        | 2251/3092 [08:14<02:52,  4.89it/s, loss=0.7189]

Step 2250 | loss : 0.7189


Epoch 2/3 [Train]:  74%|███████████████████████▊        | 2302/3092 [08:25<02:45,  4.78it/s, loss=0.5828]

Step 2300 | loss : 0.6941


Epoch 2/3 [Train]:  76%|████████████████████████▎       | 2351/3092 [08:36<02:20,  5.26it/s, loss=0.7067]

Step 2350 | loss : 0.7067


Epoch 2/3 [Train]:  78%|████████████████████████▊       | 2402/3092 [08:47<02:32,  4.53it/s, loss=0.6477]

Step 2400 | loss : 0.5186


Epoch 2/3 [Train]:  79%|█████████████████████████▎      | 2451/3092 [08:58<02:20,  4.58it/s, loss=1.0181]

Step 2450 | loss : 1.0181


Epoch 2/3 [Train]:  81%|█████████████████████████▉      | 2502/3092 [09:09<02:04,  4.74it/s, loss=0.4845]

Step 2500 | loss : 0.6469


Epoch 2/3 [Train]:  83%|██████████████████████████▍     | 2551/3092 [09:20<01:46,  5.06it/s, loss=0.5408]

Step 2550 | loss : 0.5408


Epoch 2/3 [Train]:  84%|██████████████████████████▉     | 2602/3092 [09:31<01:38,  4.99it/s, loss=0.6681]

Step 2600 | loss : 0.5605


Epoch 2/3 [Train]:  86%|███████████████████████████▍    | 2651/3092 [09:42<01:37,  4.51it/s, loss=0.4262]

Step 2650 | loss : 0.4262


Epoch 2/3 [Train]:  87%|███████████████████████████▉    | 2701/3092 [09:53<01:25,  4.58it/s, loss=0.7708]

Step 2700 | loss : 0.7708


Epoch 2/3 [Train]:  89%|████████████████████████████▍   | 2751/3092 [10:04<01:22,  4.13it/s, loss=0.7860]

Step 2750 | loss : 0.7860


Epoch 2/3 [Train]:  91%|████████████████████████████▉   | 2801/3092 [10:16<01:05,  4.41it/s, loss=0.4940]

Step 2800 | loss : 0.4940


Epoch 2/3 [Train]:  92%|█████████████████████████████▌  | 2851/3092 [10:27<00:55,  4.32it/s, loss=0.6188]

Step 2850 | loss : 0.6188


Epoch 2/3 [Train]:  94%|██████████████████████████████  | 2902/3092 [10:38<00:40,  4.67it/s, loss=0.5386]

Step 2900 | loss : 0.4531


Epoch 2/3 [Train]:  95%|██████████████████████████████▌ | 2951/3092 [10:49<00:34,  4.05it/s, loss=0.4852]

Step 2950 | loss : 0.4852


Epoch 2/3 [Train]:  97%|███████████████████████████████ | 3001/3092 [11:01<00:21,  4.15it/s, loss=0.5768]

Step 3000 | loss : 0.5768


Epoch 2/3 [Train]:  99%|███████████████████████████████▌| 3051/3092 [11:12<00:08,  4.97it/s, loss=0.5869]

Step 3050 | loss : 0.5869


Epoch 2/3 [Val]: 100%|█████████████████████████████████████████████████| 387/387 [00:24<00:00, 16.02it/s]



Bilan Epoch 2 | Loss train : 0.6844 | Val Acc : 0.5917 | Val F1 : 0.5907
  => Meilleur modèle sauvegardé (best_deberta_cmv.pt)


Epoch 3/3 [Train]:   0%|                                   | 2/3092 [00:00<10:58,  4.69it/s, loss=0.7564]

Step 0 | loss : 0.8156


Epoch 3/3 [Train]:   2%|▌                                 | 51/3092 [00:11<10:26,  4.86it/s, loss=0.5319]

Step 50 | loss : 0.5319


Epoch 3/3 [Train]:   3%|█                                | 102/3092 [00:23<10:02,  4.96it/s, loss=0.7853]

Step 100 | loss : 0.6143


Epoch 3/3 [Train]:   5%|█▌                               | 151/3092 [00:34<11:53,  4.12it/s, loss=0.8314]

Step 150 | loss : 0.8314


Epoch 3/3 [Train]:   7%|██▏                              | 201/3092 [00:45<11:47,  4.09it/s, loss=0.5296]

Step 200 | loss : 0.5296


Epoch 3/3 [Train]:   8%|██▋                              | 251/3092 [00:56<11:15,  4.21it/s, loss=0.7416]

Step 250 | loss : 0.7416


Epoch 3/3 [Train]:  10%|███▏                             | 302/3092 [01:07<09:56,  4.67it/s, loss=0.9547]

Step 300 | loss : 0.6582


Epoch 3/3 [Train]:  11%|███▋                             | 351/3092 [01:17<08:42,  5.25it/s, loss=0.6810]

Step 350 | loss : 0.6810


Epoch 3/3 [Train]:  13%|████▎                            | 401/3092 [01:29<09:42,  4.62it/s, loss=0.4958]

Step 400 | loss : 0.4958


Epoch 3/3 [Train]:  15%|████▊                            | 451/3092 [01:40<10:27,  4.21it/s, loss=0.7545]

Step 450 | loss : 0.7545


Epoch 3/3 [Train]:  16%|█████▎                           | 502/3092 [01:51<09:02,  4.77it/s, loss=0.7489]

Step 500 | loss : 0.9547


Epoch 3/3 [Train]:  18%|█████▉                           | 551/3092 [02:02<09:39,  4.39it/s, loss=0.5295]

Step 550 | loss : 0.5295


Epoch 3/3 [Train]:  19%|██████▍                          | 601/3092 [02:13<09:03,  4.58it/s, loss=0.6203]

Step 600 | loss : 0.6203


Epoch 3/3 [Train]:  21%|██████▉                          | 651/3092 [02:25<08:45,  4.64it/s, loss=0.5945]

Step 650 | loss : 0.5945


Epoch 3/3 [Train]:  23%|███████▍                         | 701/3092 [02:36<10:01,  3.97it/s, loss=0.6892]

Step 700 | loss : 0.6892


Epoch 3/3 [Train]:  24%|████████                         | 751/3092 [02:47<08:25,  4.63it/s, loss=0.4866]

Step 750 | loss : 0.4866


Epoch 3/3 [Train]:  26%|████████▌                        | 801/3092 [02:58<08:38,  4.42it/s, loss=0.5807]

Step 800 | loss : 0.5807


Epoch 3/3 [Train]:  28%|█████████                        | 851/3092 [03:10<08:22,  4.46it/s, loss=0.9617]

Step 850 | loss : 0.9617


Epoch 3/3 [Train]:  29%|█████████▌                       | 901/3092 [03:21<08:30,  4.29it/s, loss=0.8042]

Step 900 | loss : 0.8042


Epoch 3/3 [Train]:  31%|██████████▏                      | 951/3092 [03:32<08:06,  4.40it/s, loss=0.5350]

Step 950 | loss : 0.5350


Epoch 3/3 [Train]:  32%|██████████▎                     | 1001/3092 [03:43<07:48,  4.46it/s, loss=0.3715]

Step 1000 | loss : 0.3715


Epoch 3/3 [Train]:  34%|██████████▉                     | 1051/3092 [03:54<07:37,  4.46it/s, loss=0.5393]

Step 1050 | loss : 0.5393


Epoch 3/3 [Train]:  36%|███████████▍                    | 1102/3092 [04:05<06:13,  5.33it/s, loss=1.1439]

Step 1100 | loss : 0.6005


Epoch 3/3 [Train]:  37%|███████████▉                    | 1151/3092 [04:16<07:17,  4.43it/s, loss=0.9395]

Step 1150 | loss : 0.9395


Epoch 3/3 [Train]:  39%|████████████▍                   | 1202/3092 [04:27<06:41,  4.70it/s, loss=0.4153]

Step 1200 | loss : 1.0527


Epoch 3/3 [Train]:  40%|████████████▉                   | 1251/3092 [04:38<07:19,  4.19it/s, loss=0.7796]

Step 1250 | loss : 0.7796


Epoch 3/3 [Train]:  42%|█████████████▍                  | 1302/3092 [04:49<05:47,  5.15it/s, loss=0.5622]

Step 1300 | loss : 0.8849


Epoch 3/3 [Train]:  44%|█████████████▉                  | 1351/3092 [05:00<06:26,  4.51it/s, loss=0.8392]

Step 1350 | loss : 0.8392


Epoch 3/3 [Train]:  45%|██████████████▍                 | 1401/3092 [05:11<06:28,  4.35it/s, loss=0.6863]

Step 1400 | loss : 0.6863


Epoch 3/3 [Train]:  47%|███████████████                 | 1451/3092 [05:22<06:00,  4.55it/s, loss=1.0415]

Step 1450 | loss : 1.0415


Epoch 3/3 [Train]:  49%|███████████████▌                | 1502/3092 [05:33<05:14,  5.05it/s, loss=0.5184]

Step 1500 | loss : 0.8312


Epoch 3/3 [Train]:  50%|████████████████                | 1551/3092 [05:44<05:43,  4.48it/s, loss=0.6172]

Step 1550 | loss : 0.6172


Epoch 3/3 [Train]:  52%|████████████████▌               | 1601/3092 [05:55<06:20,  3.92it/s, loss=0.7902]

Step 1600 | loss : 0.7902


Epoch 3/3 [Train]:  53%|█████████████████               | 1651/3092 [06:06<05:39,  4.24it/s, loss=0.5713]

Step 1650 | loss : 0.5713


Epoch 3/3 [Train]:  55%|█████████████████▌              | 1701/3092 [06:17<05:36,  4.14it/s, loss=0.5283]

Step 1700 | loss : 0.5283


Epoch 3/3 [Train]:  57%|██████████████████              | 1751/3092 [06:28<05:04,  4.41it/s, loss=0.6907]

Step 1750 | loss : 0.6907


Epoch 3/3 [Train]:  58%|██████████████████▋             | 1801/3092 [06:38<04:20,  4.95it/s, loss=1.1550]

Step 1800 | loss : 1.1550


Epoch 3/3 [Train]:  60%|███████████████████▏            | 1851/3092 [06:49<04:23,  4.71it/s, loss=1.0454]

Step 1850 | loss : 1.0454


Epoch 3/3 [Train]:  61%|███████████████████▋            | 1901/3092 [07:00<04:45,  4.18it/s, loss=0.5574]

Step 1900 | loss : 0.5574


Epoch 3/3 [Train]:  63%|████████████████████▏           | 1951/3092 [07:11<03:49,  4.98it/s, loss=0.6652]

Step 1950 | loss : 0.6652


Epoch 3/3 [Train]:  65%|████████████████████▋           | 2001/3092 [07:22<04:17,  4.24it/s, loss=0.5121]

Step 2000 | loss : 0.5121


Epoch 3/3 [Train]:  66%|█████████████████████▏          | 2051/3092 [07:33<03:55,  4.43it/s, loss=0.4999]

Step 2050 | loss : 0.4999


Epoch 3/3 [Train]:  68%|█████████████████████▊          | 2102/3092 [07:44<03:23,  4.86it/s, loss=0.5146]

Step 2100 | loss : 0.5196


Epoch 3/3 [Train]:  70%|██████████████████████▎         | 2151/3092 [07:55<03:24,  4.60it/s, loss=0.7246]

Step 2150 | loss : 0.7246


Epoch 3/3 [Train]:  71%|██████████████████████▊         | 2201/3092 [08:07<03:33,  4.17it/s, loss=0.8210]

Step 2200 | loss : 0.8210


Epoch 3/3 [Train]:  73%|███████████████████████▎        | 2251/3092 [08:18<02:58,  4.71it/s, loss=0.6544]

Step 2250 | loss : 0.6544


Epoch 3/3 [Train]:  74%|███████████████████████▊        | 2302/3092 [08:29<02:48,  4.70it/s, loss=0.3951]

Step 2300 | loss : 0.5495


Epoch 3/3 [Train]:  76%|████████████████████████▎       | 2351/3092 [08:40<02:36,  4.74it/s, loss=0.5850]

Step 2350 | loss : 0.5850


Epoch 3/3 [Train]:  78%|████████████████████████▊       | 2402/3092 [08:51<02:14,  5.14it/s, loss=0.8006]

Step 2400 | loss : 0.9930


Epoch 3/3 [Train]:  79%|█████████████████████████▎      | 2451/3092 [09:01<02:36,  4.10it/s, loss=0.5697]

Step 2450 | loss : 0.5697


Epoch 3/3 [Train]:  81%|█████████████████████████▉      | 2501/3092 [09:13<02:12,  4.48it/s, loss=0.4280]

Step 2500 | loss : 0.4280


Epoch 3/3 [Train]:  83%|██████████████████████████▍     | 2551/3092 [09:24<01:49,  4.96it/s, loss=0.5605]

Step 2550 | loss : 0.5605


Epoch 3/3 [Train]:  84%|██████████████████████████▉     | 2602/3092 [09:35<01:35,  5.14it/s, loss=0.7347]

Step 2600 | loss : 0.9368


Epoch 3/3 [Train]:  86%|███████████████████████████▍    | 2651/3092 [09:45<01:30,  4.86it/s, loss=0.5303]

Step 2650 | loss : 0.5303


Epoch 3/3 [Train]:  87%|███████████████████████████▉    | 2702/3092 [09:57<01:16,  5.10it/s, loss=0.6389]

Step 2700 | loss : 0.6662


Epoch 3/3 [Train]:  89%|████████████████████████████▍   | 2751/3092 [10:07<01:11,  4.76it/s, loss=0.4704]

Step 2750 | loss : 0.4704


Epoch 3/3 [Train]:  91%|████████████████████████████▉   | 2801/3092 [10:19<01:10,  4.11it/s, loss=0.6605]

Step 2800 | loss : 0.6605


Epoch 3/3 [Train]:  92%|█████████████████████████████▌  | 2851/3092 [10:30<00:56,  4.28it/s, loss=0.5510]

Step 2850 | loss : 0.5510


Epoch 3/3 [Train]:  94%|██████████████████████████████  | 2902/3092 [10:41<00:41,  4.53it/s, loss=0.7544]

Step 2900 | loss : 0.4909


Epoch 3/3 [Train]:  95%|██████████████████████████████▌ | 2951/3092 [10:52<00:28,  4.90it/s, loss=0.7512]

Step 2950 | loss : 0.7512


Epoch 3/3 [Train]:  97%|███████████████████████████████ | 3002/3092 [11:03<00:19,  4.59it/s, loss=0.6627]

Step 3000 | loss : 0.6085


Epoch 3/3 [Train]:  99%|███████████████████████████████▌| 3051/3092 [11:13<00:08,  4.62it/s, loss=0.5794]

Step 3050 | loss : 0.5794


Epoch 3/3 [Val]: 100%|█████████████████████████████████████████████████| 387/387 [00:24<00:00, 16.08it/s]


Bilan Epoch 3 | Loss train : 0.6746 | Val Acc : 0.5762 | Val F1 : 0.5752


In [71]:
print("\nChargement du meilleur modèle pour l'évaluation finale...")
model_deb.load_state_dict(torch.load('best_deberta_cmv.pt'))
model_deb.eval()
 
test_preds, test_labels_list = [], []
 
with torch.no_grad():
    for batch in tqdm(test_loader_deb, desc="Évaluation Test"):
        input_ids      = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
 
        outputs = model_deb(input_ids=input_ids, attention_mask=attention_mask)
        preds = torch.argmax(outputs.logits, dim=1)
        test_preds.extend(preds.cpu().numpy())
        test_labels_list.extend(labels.cpu().numpy())
 
final_accuracy = accuracy_score(test_labels_list, test_preds)
final_f1       = f1_score(test_labels_list, test_preds, average='macro')
 
print("\n" + "="*50)
print("RÉSULTATS FINAUX — DEBERTA-V3-LARGE SUR LE TEST")
print("="*50)
print(f"Précision (Accuracy) : {final_accuracy:.4f}")
print(f"Score F1  (Macro)    : {final_f1:.4f}")
print()
print(classification_report(
    test_labels_list,
    test_preds,
    target_names=["Option 0 Gagne", "Option 1 Gagne"]
))



Chargement du meilleur modèle pour l'évaluation finale...


Évaluation Test: 100%|█████████████████████████████████████████████████| 387/387 [00:23<00:00, 16.37it/s]


RÉSULTATS FINAUX — DEBERTA-V3-LARGE SUR LE TEST
Précision (Accuracy) : 0.6486
Score F1  (Macro)    : 0.3934

                precision    recall  f1-score   support

Option 0 Gagne       1.00      0.65      0.79       387
Option 1 Gagne       0.00      0.00      0.00         0

      accuracy                           0.65       387
     macro avg       0.50      0.32      0.39       387
  weighted avg       1.00      0.65      0.79       387

